pkg/apiserver/slowquery/queries.go

```diff
@@ -117,84 +123,42 @@ type GetListRequest struct {
 	Fields string `json:"fields" form:"fields"` // example: "Query,Digest"
 }
 
-var cachedProjectionsMap map[string]string
-
-func getProjectionsMap() map[string]string {
-	if cachedProjectionsMap == nil {
-		t := reflect.TypeOf(SlowQuery{})
-		fieldsNum := t.NumField()
-		ret := map[string]string{}
-		for i := 0; i < fieldsNum; i++ {
-			field := t.Field(i)
-			// ignore to check error because the field is defined by ourself
-			// we can confirm that it has "gorm" tag and fixed structure
-			s, _ := field.Tag.Lookup("gorm")
-			jsonField := strings.ToLower(field.Tag.Get("json"))
-			sourceField := strings.Split(s, ":")[1]
-			if proj, ok := field.Tag.Lookup("proj"); ok {
-				ret[jsonField] = fmt.Sprintf("%s AS %s", proj, sourceField)
-			} else {
-				ret[jsonField] = sourceField
-			}
-		}
-		cachedProjectionsMap = ret
-	}
-	return cachedProjectionsMap
-}
-
-func getProjectionsByFields(jsonFields ...string) ([]string, error) {
-	projMap := getProjectionsMap()
-	ret := make([]string, 0, len(jsonFields))
-	for _, fieldName := range jsonFields {
-		field, ok := projMap[strings.ToLower(fieldName)]
-		if !ok {
-			return nil, fmt.Errorf("unknown field %s", fieldName)
-		}
-		ret = append(ret, field)
-	}
-	return ret, nil
-}
-
-var cachedAllProjections []string
-
-func getAllProjections() []string {
-	if cachedAllProjections == nil {
-		projMap := getProjectionsMap()
-		ret := make([]string, 0, len(projMap))
-		for _, proj := range projMap {
-			ret = append(ret, proj)
-		}
-		cachedAllProjections = ret
-	}
-	return cachedAllProjections
-}
-
 type GetDetailRequest struct {
 	Digest    string  `json:"digest" form:"digest"`
 	Timestamp float64 `json:"timestamp" form:"timestamp"`
 	// TODO: Switch back to uint64 when modern browser as well as Swagger handles BigInt well.
 	ConnectID string `json:"connect_id" form:"connect_id"`
 }
 
-func QuerySlowLogList(db *gorm.DB, req *GetListRequest) ([]SlowQuery, error) {
-	var projections []string
+func (s *Service) querySlowLogList(db *gorm.DB, req *GetListRequest) ([]SlowQuery, error) {
+	var fields []string
 	var err error
 	reqFields := strings.Split(req.Fields, ",")
+	if err != nil {
+		return nil, err
+	}
+
+	tableColumns, err := s.params.SysSchema.GetTableColumnNames(db, slowQueryTable)
+	if err != nil {
+		return nil, err
+	}
+
 	if len(reqFields) == 1 && reqFields[0] == "*" {
-		projections = getAllProjections()
+		fields, err = filterFieldsBy(SlowQuery{}, tableColumns)
 	} else {
-		projections, err = getProjectionsByFields(
+		mustContainsFields := []string{"digest", "connection_id", "timestamp"}
+		fields, err = filterFieldsBy(SlowQuery{}, tableColumns,
 			funk.UniqString(
-				append([]string{"digest", "connection_id", "timestamp"}, reqFields...),
+				append(mustContainsFields, reqFields...),
 			)...)
```

Comment:
How about (pseudo code):

```golang
columns := (requiredFields ∪ reqFields) ∩ tableColumns
filterFieldsBy(SlowQuery{}, columns)
```

Then the logic of your `filterFieldsBy` can be simplified.